test/test_foreach.py

```diff
@@ -981,6 +983,34 @@ def inplace_func(*tensorlist):
                     check_batched_grad=False,
                 )
 
+    @unittest.skipIf(not (torch.cuda.is_available() and torch.cuda.device_count() > 1), "requires multiple GPUs")
+    def test_tensors_grouping(self):
+        num_tensors_per_list = 10
+        num_devices = torch.cuda.device_count()
+        dtypes = (torch.float16, torch.float32, torch.float64)
+        list1 = [
+            torch.tensor(
+                i,
+                device=torch.device("cuda", random.randint(0, num_devices - 1)),
+                dtype=dtypes[random.randint(0, 2)],
+            ) for i in range(num_tensors_per_list)
+        ]
+        list2 = [None for _ in list1]
+        list3 = [torch.rand_like(t) for t in list1]
+        nested_tensorlists = [list1, list2, list3]
+        grouped_tensors = torch.utils._foreach_utils._group_tensors_by_device_and_dtype(nested_tensorlists, with_indices=True)
+        num_tensors_seen = 0
+        for (device, dtype), ([l1, l2, l3], indices) in grouped_tensors.items():
+            for t in itertools.chain(l1, l3):
+                self.assertEquals(t.device, device)
+                self.assertEquals(t.dtype, dtype)
+                num_tensors_seen += 1
```

Comment:
```suggestion
                num_tensors_seen += 1
            assert len(l1) == len(l2) and all(p == None for p in l2)
```
I know the following code implicitly checks, but this would be a stricter check (since we never verify the length of l2 below.